x-pack/plugins/security_solution/common/api/detection_engine/prebuilt_rules/model/diff/diffable_rule/diffable_rule.ts

```diff
@@ -207,21 +207,79 @@ const DiffableRule = z.intersection(
+  DiffableThresholdFields.shape.type,
+  DiffableMachineLearningFields.shape.type,
+  DiffableNewTermsFields.shape.type,
+]);
+
 /**
  * This is a merge of all fields from all rule types into a single TS type.
  * This is NOT a union discriminated by rule type, as DiffableRule is.
  */
-export type DiffableAllFields = DiffableCommonFields &
-  Omit<DiffableCustomQueryFields, 'type'> &
-  Omit<DiffableSavedQueryFields, 'type'> &
-  Omit<DiffableEqlFields, 'type'> &
-  Omit<DiffableEsqlFields, 'type'> &
-  Omit<DiffableThreatMatchFields, 'type'> &
-  Omit<DiffableThresholdFields, 'type'> &
-  Omit<DiffableMachineLearningFields, 'type'> &
-  Omit<DiffableNewTermsFields, 'type'> &
-  DiffableRuleTypeField;
-
-interface DiffableRuleTypeField {
-  type: DiffableRule['type'];
-}
+// export type DiffableAllFields = DiffableCommonFields &
+//   Omit<DiffableCustomQueryFields, 'type'> &
+//   Omit<DiffableSavedQueryFields, 'type'> &
+//   Omit<DiffableEqlFields, 'type'> &
+//   Omit<DiffableEsqlFields, 'type'> &
+//   Omit<DiffableThreatMatchFields, 'type'> &
+//   Omit<DiffableThresholdFields, 'type'> &
+//   Omit<DiffableMachineLearningFields, 'type'> &
+//   Omit<DiffableNewTermsFields, 'type'> &
+//   DiffableRuleTypeField;
+
+type DiffableAllFields = z.infer<typeof DiffableAllFields>;
+const DiffableAllFields = DiffableCommonFields.merge(DiffableCustomQueryFields.omit({ type: true }))
+  .merge(DiffableSavedQueryFields.omit({ type: true }))
+  .merge(DiffableEqlFields.omit({ type: true }))
+  .merge(DiffableEsqlFields.omit({ type: true }))
+  .merge(DiffableThreatMatchFields.omit({ type: true }))
+  .merge(DiffableThresholdFields.omit({ type: true }))
+  .merge(DiffableMachineLearningFields.omit({ type: true }))
+  .merge(DiffableNewTermsFields.omit({ type: true }))
+  .merge(z.object({ type: DiffableRuleTypes }));
+
+type DiffableUpgradableFields = z.infer<typeof DiffableUpgradableFields>;
+const DiffableUpgradableFields = DiffableAllFields.omit({
+  type: true,
+  rule_id: true,
+  version: true,
+});
+
+export type PickVersionValues = z.infer<typeof PickVersionValues>;
+export const PickVersionValues = z.enum(['BASE', 'CURRENT', 'TARGET', 'MERGED']);
+
+type FieldUpgradeSpecifier<T> = z.infer<ReturnType<typeof fieldUpgradeSpecifier<z.ZodType<T>>>>;
+const fieldUpgradeSpecifier = <T extends z.ZodTypeAny>(fieldSchema: T) =>
+  z.discriminatedUnion('pick_version', [
+    z.object({
+      pick_version: PickVersionValues,
+    }),
+    z.object({
+      pick_version: z.literal('RESOLVED'),
+      resolved_value: fieldSchema,
+    }),
+  ]);
+
+type FieldUpgradeSpecifiers<TFields> = Required<{
+  [Field in keyof TFields]: FieldUpgradeSpecifier<TFields[Field]>;
+}>;
+
+type RuleFieldsToUpgrade = FieldUpgradeSpecifiers<DiffableUpgradableFields>;
+const RuleFieldsToUpgrade = z
+  .object(
+    mapValues(DiffableUpgradableFields.shape, (fieldSchema) => {
+      return fieldUpgradeSpecifier(fieldSchema);
+    })
+  )
+  .strict();
```

Comment:
@jpdjere I think we forgot that the fields in `RuleFieldsToUpgrade` should be all optional

```suggestion
type FieldUpgradeSpecifiers<TFields> = {
  [Field in keyof TFields]?: FieldUpgradeSpecifier<TFields[Field]>;
};

type RuleFieldsToUpgrade = FieldUpgradeSpecifiers<DiffableUpgradableFields>;
const RuleFieldsToUpgrade = z
  .object(
    mapValues(DiffableUpgradableFields.shape, (fieldSchema) => {
      return fieldUpgradeSpecifier(fieldSchema).optional();
    })
  )
  .strict();
```

Reply:
Thanks, fixed